src/js_native_api_v8.cc

```diff
@@ -302,30 +312,34 @@ class Reference : private Finalizer {
     data.SetSecondPassCallback(SecondPassCallback);
   }
 
-  static void SecondPassCallback(const v8::WeakCallbackInfo<Reference>& data) {
-    Reference* reference = data.GetParameter();
-
-    napi_env env = reference->_env;
-
-    if (reference->_finalize_callback != nullptr) {
+  inline void RunFinalizer() {
+    napi_env env = _env;
+    if (_finalize_callback != nullptr) {
       NapiCallIntoModuleThrow(env, [&]() {
-        reference->_finalize_callback(
-            reference->_env,
-            reference->_finalize_data,
-            reference->_finalize_hint);
+        _finalize_callback(_env, _finalize_data, _finalize_hint);
       });
     }
 
-    // this is safe because if a request to delete the reference
-    // is made in the finalize_callback it will defer deletion
-    // to this block and set _delete_self to true
-    if (reference->_delete_self) {
-      Delete(reference);
+    // This is safe because if a request to delete the reference is made in the
+    // finalize_callback it will defer deletion to this block and set
+    // _delete_self to true.
+    if (_delete_self) {
+      Delete(this);
     } else {
-      reference->_finalize_ran = true;
+      _finalize_ran = true;
     }
   }
 
+  static void SecondPassCallback(const v8::WeakCallbackInfo<Reference>& data) {
+    data.GetParameter()->RunFinalizer();
+  }
+
+  static void EnvCleanupHook(void* data) {
+    Reference* reference = static_cast<Reference*>(data);
+    reference->_persistent.Reset();
+    reference->RunFinalizer();
```

Comment:
Is it expected that when the EnvCleanupHook is being run that it will also remove itself from the list of cleanup hooks? The current code does this (since it is necessary if the finalizer is run normally).  If nothing else it may be extra work that is not necessary.